precompiles/staking/tx_test.go

```diff
@@ -14,6 +15,299 @@ import (
+			true,
+			"invalid type for",
+		},
+		{
+			"fail - invalid pubkey",
+			func() []interface{} {
+				return []interface{}{
+					description,
+					commission,
+					minSelfDelegation,
+					delegatorAddress,
+					validatorAddress,
+					1205,
+					value,
+				}
+			},
+			200000,
+			func(data []byte) {},
+			true,
+			"invalid type for",
+		},
+		{
+			"fail - invalid pubkey",
+			func() []interface{} {
+				return []interface{}{
+					description,
+					commission,
+					minSelfDelegation,
+					delegatorAddress,
+					validatorAddress,
+					"bHVrZQ=", // base64.StdEncoding.DecodeString error
+					value,
+				}
+			},
+			200000,
+			func(data []byte) {},
+			true,
+			"",
+		},
+		{
+			"fail - invalid value",
+			func() []interface{} {
+				return []interface{}{
+					description,
+					commission,
+					minSelfDelegation,
+					delegatorAddress,
+					validatorAddress,
+					pubkey,
+					"",
+				}
+			},
+			200000,
+			func(data []byte) {},
+			true,
+			"invalid amount",
+		},
+		{
+			"success",
+			func() []interface{} {
+				return []interface{}{
+					description,
+					commission,
+					minSelfDelegation,
+					delegatorAddress,
+					validatorAddress,
+					pubkey,
+					value,
+				}
+			},
+			200000,
+			func(data []byte) {
+				success, err := s.precompile.Unpack(staking.CreateValidatorMethod, data)
+				s.Require().NoError(err)
+				s.Require().Equal(success[0], true)
+
+				log := s.stateDB.Logs()[0]
+				s.Require().Equal(log.Address, s.precompile.Address())
+
+				// Check event signature matches the one emitted
+				event := s.precompile.ABI.Events[staking.EventTypeCreateValidator]
+				s.Require().Equal(crypto.Keccak256Hash([]byte(event.Sig)), geth.HexToHash(log.Topics[0].Hex()))
+				s.Require().Equal(log.BlockNumber, uint64(s.ctx.BlockHeight()))
+
+				// Check the fully unpacked event matches the one emitted
+				var createValidatorEvent staking.EventCreateValidator
+				err = cmn.UnpackLog(s.precompile.ABI, &createValidatorEvent, staking.EventTypeCreateValidator, *log)
+				s.Require().NoError(err)
+				s.Require().Equal(s.address, createValidatorEvent.DelegatorAddress)
+				s.Require().Equal(s.address, createValidatorEvent.ValidatorAddress)
+				s.Require().Equal(value, createValidatorEvent.Value)
+			},
+			false,
+			"",
+		},
+	}
+
+	for _, tc := range testCases {
+		s.Run(tc.name, func() {
+			s.SetupTest()
+
+			// reset sender
+			delegatorAddress = s.address
+			validatorAddress = sdk.ValAddress(s.address.Bytes()).String()
+
+			var contract *vm.Contract
+			contract, s.ctx = testutil.NewPrecompileContract(s.T(), s.ctx, s.address, s.precompile, tc.gas)
+
+			bz, err := s.precompile.CreateValidator(s.ctx, s.address, contract, s.stateDB, &method, tc.malleate())
+
+			// query the validator in the staking keeper
+			validator := s.app.StakingKeeper.Validator(s.ctx, s.address.Bytes())
+			if tc.expError {
+				s.Require().ErrorContains(err, tc.errContains)
+				s.Require().Empty(bz)
```

Comment:
In the error case one could also check that the returned validator is `nil`
```suggestion
			// query the validator in the staking keeper
			validator := s.app.StakingKeeper.Validator(s.ctx, s.address.Bytes())
			if tc.expError {
				s.Require().ErrorContains(err, tc.errContains)
				s.Require().Empty(bz)
				s.Require().Nil(validator)
```

Reply:
done